File: app/src/main/java/space/pal/sig/repository/client/BaseClient.java
package space.pal.sig.repository.client;

import com.google.gson.Gson;

import lombok.RequiredArgsConstructor;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * SpaceNow
 * Created by Catalin on 7/19/2020
 **/
@RequiredArgsConstructor
public abstract class BaseClient {

    private static Retrofit CLIENT;
    private final OkHttpClient okHttpClient;
    private final Gson gson;

    public <T> T createService(final Class<T> service) {
        return getClient().create(service);
    }

    public abstract String getApiUrl();

    private Retrofit getClient() {
        if (CLIENT == null) {
            synchronized (BaseClient.class) {
                CLIENT = new Retrofit.Builder()
                        .baseUrl(getApiUrl())
                        .addConverterFactory(GsonConverterFactory.create(gson))
                        .client(okHttpClient)
                        .build();
            }
        }
        return CLIENT;
    }

}

File: app/src/main/java/space/pal/sig/network/client/BaseClient.kt
package space.pal.sig.network.client

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

/**
SpaceNow
Created by Catalin on 12/6/2020
 **/
abstract class BaseClient {

    private var client: Retrofit? = null

    fun <T> createService(service: Class<T>): T {
        return getClient().create(service)
    }

    abstract fun getApiUrl(): String

    private fun getClient(): Retrofit {
        if (client == null) {
            synchronized(BaseClient::class.java) {
                client = build(getApiUrl())
            }
        }
        return client!!
    }

    private fun build(url: String): Retrofit {
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        val okHttpClient = OkHttpClient
                .Builder()
                .addInterceptor(loggingInterceptor)
                .build()
        return Retrofit.Builder()
                .baseUrl(url)
                .client(okHttpClient)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .build()
    }

}